lib/tumblr.js

```diff
@@ -1,262 +1,452 @@
+    delete currentRequest.body;
 
-  reblog: function (blogName, options, callback) {
-    this._post(blogPath(blogName, '/post/reblog'), options, callback);
-  },
+    // And then add the full body
+    var form = currentRequest.form();
+    for (var key in params) {
+        form.append(key, params[key]);
+    }
 
-  deletePost: function (blogName, id, callback) {
-    this._post(blogPath(blogName, '/post/delete'), {id: id}, callback);
-  },
+    // Add the form header back
+    extend(currentRequest.headers, form.getHeaders());
 
-  photo: calls.postCreation('photo', ['data', 'source'],       true),
-  audio: calls.postCreation('audio', ['data', 'external_url'], true),
-  video: calls.postCreation('video', ['data', 'embed'],        true),
-  quote: calls.postCreation('quote', ['quote'],                false),
-  text:  calls.postCreation('text',  ['body'],                 false),
-  link:  calls.postCreation('link',  ['url'],                  false),
-  chat:  calls.postCreation('chat',  ['conversation'],         false),
+    return currentRequest;
+}
 
-  // User
+/**
+ * ## addMethod
+ *
+ * Adds a request method to the client
+ *
+ * @param  {Object} client: add the method to this object
+ * @param  {String} methodName: the name of the method
+ * @param  {String} apiPath: the API route, which uses any colon-prefixed segments as arguments
+ * @param  {Array} paramNames: ordered list of required request parameters used as arguments
+ * @param  {String|Function} requestType: the request type or a function that makes the request
+ */
+function addMethod(client, methodName, apiPath, paramNames, requestType) {
+    var apiPathSplit = apiPath.split('/');
+    var apiPathParamsCount = apiPath.split(/\/:[^\/]+/).length - 1;
+
+    var buildApiPath = function(args) {
+        var pathParamIndex = 0;
+        return reduce(apiPathSplit, function(apiPath, apiPathChunk, i) {
+            // Parse arguments in the path
+            if (apiPathChunk === ':blogIdentifier') {
+                // Blog URLs are special
+                apiPathChunk = forceFullBlogUrl(args[pathParamIndex++]);
+            } else if (apiPathChunk[0] === ':') {
+                apiPathChunk = args[pathParamIndex++];
+            }
+
+            if (apiPathChunk) {
+                return apiPath + '/' + apiPathChunk;
+            } else {
+                return apiPath;
+            }
+        }, '');
+    };
 
-  userInfo: function (callback) {
-    this._get('/user/info', {}, callback);
-  },
+    var namedParams = (apiPath.match(/\/:[^\/]+/g) || []).map(function(param) {
+        return param.substr(2);
+    }).concat(paramNames, 'params', 'callback');
 
-  likes: function (options, callback) {
-    if (isFunction(options)) { callback = options; options = {}; }
-    this._get('/user/likes', options, callback);
-  },
+    var methodBody = function() {
+        var argsLength = arguments.length;
+        var args = new Array(argsLength);
```

Comment:
Probably doesn't have a huge performance impact either way in this case, but using `slice` on `arguments` makes the function un-optimizable in V8: https://github.com/petkaantonov/bluebird/wiki/Optimization-killers#3-managing-arguments